src/GafferScene/CollectScenes.cpp

```diff
@@ -0,0 +1,430 @@
+
+using namespace std;
+using namespace Imath;
+using namespace IECore;
+using namespace Gaffer;
+using namespace GafferScene;
+
+//////////////////////////////////////////////////////////////////////////
+// Internal utilities
+//////////////////////////////////////////////////////////////////////////
+
+namespace
+{
+
+class SceneScope : public Context::EditableScope
+{
+
+	public :
+
+		SceneScope( const Context *context, const InternedString &rootNameVariable )
+			:	EditableScope( context ), m_rootNameVariable( rootNameVariable )
+		{
+		}
+
+		SceneScope( const Context *context, const InternedString &rootNameVariable, const ScenePlug::ScenePath &downstreamPath )
+			:	EditableScope( context ), m_rootNameVariable( rootNameVariable )
+		{
+			setScenePath( downstreamPath );
+		}
+
+		void setRootName( const InternedString &name )
+		{
+			set( m_rootNameVariable, name.string() );
+		}
+
+		void setScenePath( const ScenePlug::ScenePath &downstreamPath )
+		{
+			setRootName( downstreamPath[0] );
+			ScenePlug::ScenePath upstreamPath( downstreamPath.begin() + 1, downstreamPath.end() );
+			set( ScenePlug::scenePathContextName, upstreamPath );
+		}
+
+	private :
+
+		InternedString m_rootNameVariable;
+
+};
+
+} // namespace
+
+//////////////////////////////////////////////////////////////////////////
+// CollectScenes
+//////////////////////////////////////////////////////////////////////////
+
+IE_CORE_DEFINERUNTIMETYPED( CollectScenes );
+
+size_t CollectScenes::g_firstPlugIndex = 0;
+
+CollectScenes::CollectScenes( const std::string &name )
+	:	SceneProcessor( name, 1 )
+{
+	storeIndexOfNextChild( g_firstPlugIndex );
+
+	addChild( new StringVectorDataPlug( "rootNames", Plug::In, new StringVectorData() ) );
+	addChild( new StringPlug( "rootNameVariable", Plug::In, "collect:rootName" ) );
+}
+
+CollectScenes::~CollectScenes()
+{
+}
+
+Gaffer::StringVectorDataPlug *CollectScenes::rootNamesPlug()
+{
+	return getChild<StringVectorDataPlug>( g_firstPlugIndex + 0 );
+}
+
+const Gaffer::StringVectorDataPlug *CollectScenes::rootNamesPlug() const
+{
+	return getChild<StringVectorDataPlug>( g_firstPlugIndex + 0 );
+}
+
+Gaffer::StringPlug *CollectScenes::rootNameVariablePlug()
+{
+	return getChild<StringPlug>( g_firstPlugIndex + 1 );
+}
+
+const Gaffer::StringPlug *CollectScenes::rootNameVariablePlug() const
+{
+	return getChild<StringPlug>( g_firstPlugIndex + 1 );
+}
+
+void CollectScenes::affects( const Gaffer::Plug *input, AffectedPlugsContainer &outputs ) const
+{
+	SceneProcessor::affects( input, outputs );
+
+	if( input == rootNamesPlug() )
+	{
+		outputs.push_back( outPlug()->childNamesPlug() );
+	}
+	else if( input == rootNameVariablePlug() )
+	{
+		for( PlugIterator it( outPlug() ); !it.done(); ++it )
+		{
+			outputs.push_back( it->get() );
+		}
+	}
+	else if( input->parent<ScenePlug>() == inPlug() )
+	{
+		outputs.push_back( outPlug()->getChild<Plug>( input->getName() ) );
+	}
+	else if( input == outPlug()->childNamesPlug() )
```

Comment:
> Out of curiosity - what's the difference between having this block here instead of putting its contents into the block where the rootNamesPlug() is handled? 

The key rule is that the `affects()` method has to match what the `compute()` methods are going to do. In this case, the computes for globals/setNames/set all pull on `outPlug()->childNamesPlug()` rather than on `rootNamesPlug()`, so that is what we need to declare in `affects()`.

In practice, to the outside observer the effects would be pretty similar if we took a shortcut and dirtied directly in the earlier block, and we could probably get away with it. It would just be harder to verify correctness, be misleading to things like DownstreamIterator, and be vulnerable to changes in the implementation of the node if we add further plugs that affect `outPlug()->childNamesPlug()`.

> Is that only because we don't want to trigger a recompute if the plug value changes from 'red green blue' to 'red green blue blue'?

No, the `affects()` method is used to propagate dirtiness through the graph, and at that point we don't consider plug values at all. Dirty propagation is just to notify observers that "hey, something you looked at before has changed". It's then up to the observer to pull any updates it wants by calling `ValuePlug::getValue()`. At that point `hash()` is used to compute the cache key, and then `compute()` is used to compute a fresh result if the cache doesn't contain what we need. So, it's actually `hash()` that is responsible for reusing results. By using the already deduplicated names from `outPlug()->childNames()` in spots like `hashSetNames()` we do get that reuse for the "red green blue blue" case you described, but not because of `affects()`.